packages/contracts/contracts/maci/Tally.sol

```diff
@@ -0,0 +1,321 @@
+  uint256 public voiceCreditFactor;
+
+  /// @notice The cooldown duration for withdrawal extra funds
+  uint256 public cooldown;
+
+  /// @notice The sum of tally result squares
+  uint256 public totalVotesSquares;
+
+  /// @notice Fixed recipient count for claim
+  uint256 public recipientCount;
+
+  /// @notice Initialized or not
+  bool internal initialized;
+
+  /// @notice custom errors
+  error CooldownPeriodNotOver();
+  error VotingPeriodNotOver();
+  error VotingPeriodOver();
+  error InvalidBudget();
+  error NoProjectHasMoreThanOneVote();
+  error InvalidWithdrawal();
+  error AlreadyInitialized();
+  error NotInitialized();
+  error InvalidPoll();
+  error NotCompletedResults();
+  error TooManyResults();
+
+  /// @notice Create a new Tally contract
+  /// @param verifierContract The Verifier contract
+  /// @param vkRegistryContract The VkRegistry contract
+  /// @param pollContract The Poll contract
+  /// @param mpContract The MessageProcessor contract
+  /// @param tallyOwner The owner of the Tally contract
+  /// @param pollMode The mode of the poll
+  constructor(
+    address verifierContract,
+    address vkRegistryContract,
+    address pollContract,
+    address mpContract,
+    address tallyOwner,
+    Mode pollMode
+  ) payable TallyBase(verifierContract, vkRegistryContract, pollContract, mpContract, tallyOwner, pollMode) {}
+
+  /// @notice A modifier that causes the function to revert if the cooldown period is not over
+  modifier afterCooldown() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration + cooldown) {
+      revert CooldownPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is over
+  modifier beforeVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed > duration) {
+      revert VotingPeriodOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the voting period is not over
+  modifier afterVotingDeadline() {
+    (uint256 deployTime, uint256 duration) = poll.getDeployTimeAndDuration();
+    uint256 secondsPassed = block.timestamp - deployTime;
+
+    if (secondsPassed <= duration) {
+      revert VotingPeriodNotOver();
+    }
+
+    _;
+  }
+
+  /// @notice A modifier that causes the function to revert if the strategy is not initialized
+  modifier isInitialized() {
+    if (!initialized) {
+      revert NotInitialized();
+    }
+
+    _;
+  }
+
+  /// @notice Initialize tally with strategy params
+  /// @param params The strategy initialization params
+  function init(IPayoutStrategy.StrategyInit calldata params) public onlyOwner {
+    if (initialized) {
+      revert AlreadyInitialized();
+    }
+
+    if (params.poll != address(poll)) {
```

Comment:
why poll here? it's already part of the constructor args

Reply:
I'll try to check if it's possible to use it, but I added this check because poll can be without defined registry. This logic was in constructor before, so I think poll can be removed from init params.